### PR TITLE
API-21315: Token Parameter processing part 2

### PR DESCRIPTION
**Pull request by jabando-va**

### CODE REVIEW

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 1]
```diff
@@ -10,41 +10,53 @@
...
         .build();
   }
 
-  public boolean systemExistsIn(String... systems) {
+  public boolean isSystemOneOf(String... systems) {
     return Arrays.stream(systems).anyMatch(s -> s.equals(system()));
   }
 
   @Override
   public String toString() {
```
[the remark this answers is not in the document]
Comment (jhulbert-va) [reply fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:49]: I think we may need an enum that represents where the pipe should (or shouldn't be)

[the remark this answers is not in the document]
Comment (jabando-va) [reply fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:49]: parseToken populates `original`, so its toString should give back `|code`

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 2]
```diff
@@ -9,42 +9,50 @@
...
-    return originalString();
+    if (isBlank(system())) {
+      return code();
+    } else if (EXPLICITLY_NONE.equals(system())) {
+      return "|" + code();
+    }
+    if (isBlank(code())) {
+      return system();
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:51]: I think code should be `@NonNull`. If you look at the [spec for token searching](https://www.hl7.org/fhir/search.html#token) code will always either have a value or be `EXPLICITLY_NONE`. You can't have `system` only `system|` without a code.

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 3]
```diff
@@ -9,42 +9,50 @@
...
-  public boolean systemExistsIn(String... systems) {
+  public boolean isSystemOneOf(String... systems) {
     return Arrays.stream(systems).anyMatch(s -> s.equals(system()));
   }
 
   @Override
   public String toString() {
-    return originalString();
+    if (isBlank(system())) {
+      return code();
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:46]: You'll need to make sure `code` isn't `EXPLICITLY_NONE` in this case.

File: fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java [hunk 4]
```diff
@@ -4,47 +4,71 @@
...
+  public boolean isSystemOneOf(String... systems) {
     return Arrays.stream(systems).anyMatch(s -> s.equals(system()));
   }
 
   @Override
   public String toString() {
-    return originalString();
+    // Must have a non-blank code
+    if (isBlank(code())) {
+      throw new InvalidRequest("|");
```
Comment (jhulbert-va) [inline fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:56]: I think we will want these exceptions to be better.

Comment (jabando-va) [reply fhir-common/src/main/java/gov/va/api/health/fhir/api/FhirTokenParameter.java:56]: InvalidToken?

**Review by jhulbert-va — APPROVED** [review]

**Review by jblefkove-va — APPROVED** [review]

**Pull request merged**